File: readability.py
import cs50

text = cs50.get_string("Text: ")

# since we count spaces for words, init words to 1 instead of 0
characters = 0
words = 1
sentences = 0

# check all characters and add corresponding values for spaces = words, punctuation = sentences...
for char in text:
    if char.isalpha() == True:
        characters += 1
    elif char == " ":
        words += 1
    elif char == "." or char == "?" or char == "!":
        sentences += 1

# average number of letters per 100 words
L = characters * 100.0 / words

# average number of sentences per 100 words
S = sentences * 100.0 / words

# Coleman-Liau index
total = round(0.0588 * L - 0.296 * S - 15.8)

# convert total to the text output (if less than 1 or greater than 16, etc.)
if total < 1.0:
    grade = "Before Grade 1"
elif total >= 16.0:
    grade = "Grade 16+"
else:
    grade = f"Grade {round(total)}"

print(grade)

File: readability.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <math.h>

int count_chars();
int count_sentences();
int count_words();

int main(void)
{
    // Request text from user
    string text = get_string("Text: ");

    // Count the number of chars in string
    int chars = count_chars(text);

    // Check each character for a space (counts word)
    int words = count_words(text);

    // Check each character for period (counts sentences)
    int sentences = count_sentences(text);

    // Average letters per 100 words
    float L = chars * 100.0 / words;


    // Average sentences per 100 words
    float S = sentences * 100.0 / words;

    // Calculate Coleman-Liau index
    float index = 0.0588 * L - 0.296 * S - 15.8;

    // Round up/down the index to nearest int
    index = round(index);

    // Print under 1 grade or over 16 grade, if applies
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", (int) index);
    }
}

int count_chars(string word)
{
    int chars = 0;

    // Count all alphabetic characters
    for (int i = 0, n = strlen(word); i < n; i++)
    {
        if (word[i] >= 65 && word[i] <= 90)
        {
            chars += 1;
        }
        else if (word[i] >= 97 && word[i] <= 122)
        {
            chars += 1;
        }
    }

    return chars;
}

int count_sentences(string word)
{
    int sentences = 0;

    // If characters '.', '!' or '?' are identified, count as end of a sentence
    for (int i = 0, n = strlen(word); i < n; i++)
    {
        if (word[i] == 46 || word[i] == 33 || word [i] == 63)
        {
            sentences += 1;
        }
    }

    return sentences;
}

int count_words(string word)
{
    // Since there are always n + 1 words, where n = spaces, we add 1 to the variable when calling it initially
    int words = 1;

    // Any spaces are counted as words
    for (int i = 0, n = strlen(word); i < n; i++)
    {
        if (word[i] == 32)
        {
            words += 1;
        }
    }

    return words;
}